drake/systems/frames/CoordinateFrame.h

```diff
@@ -0,0 +1,134 @@
+#ifndef __CoordinateFrame_H__
+#define __CoordinateFrame_H__
+
+#include <iostream>
+#include <string>
+#include <memory>
+#include <vector>
+#include <stdexcept>
+#include <initializer_list>
+
+#undef DLLEXPORT
+#if defined(WIN32) || defined(WIN64)
+#if defined(drakeSystem_EXPORTS)
+#define DLLEXPORT __declspec( dllexport )
+#else
+#define DLLEXPORT __declspec( dllimport )
+#endif
+#else
+#define DLLEXPORT
+#endif
+
+class DrakeSystem;
+typedef std::shared_ptr<DrakeSystem> DrakeSystemPtr;
+
+/// Every input, state, and output in a DynamicalSystem has a coordinate frame
+/// attached to it.  Many bugs can be avoided by forcing developers to be
+/// explicit about these coordinate systems when they make combinations of systems.
+
+class DLLEXPORT CoordinateFrame {
+public:
+  CoordinateFrame(const std::string& _name, const std::vector<std::string>& _coordinates)
+          : name(_name), coordinates(_coordinates) {};
+  CoordinateFrame(const std::string& _name, const unsigned int dim, const std::string& prefix)
+          : name(_name) {
+    for (int i=0; i<dim; i++) coordinates.push_back(prefix+std::to_string(i));
+  }
+  CoordinateFrame(const std::string& _name) : name(_name) {}
+
+  virtual ~CoordinateFrame(void) {  };
+
+  virtual unsigned int getDim() const { return coordinates.size(); };
+  virtual const std::string& getCoordinateName(unsigned int i) const {
+    if (i>=coordinates.size())
+      throw std::runtime_error("index exceeds dimension of coordinate frame");
+    return coordinates[i];
+  }
+  virtual std::vector<std::string> getCoordinateNames() const { return coordinates; }
+  virtual void setCoordinateNames(const std::vector<std::string>& _coordinates)
+  {
+    if (_coordinates.size()!=coordinates.size())
+      throw std::runtime_error("(CoordinateFrame:setCoordinateNames) this would change the dimension of the frame");  // which would, e.g. confuse a multi-coordinate frame that was pointing to this
+    coordinates = _coordinates;
+  }
+
+  virtual std::ostream& print(std::ostream& os) const {
+    os << "Coordinate Frame: " << name << " (" << getDim() << " elements)" << std::endl;
+
+    for (auto c : coordinates) {
```

Comment:
by the way (not critical in this use case), this does create a copy of every string (i.e. c will be of type string). You can avoid the copy by using
```
for (const auto& c : coordinates)
```